File: usr/share/installer/installer.c
#include <gtk/gtk.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

void ejecutar_comando(const char *cmd) {
    int ret = system(cmd);
    if (ret != 0) {
        fprintf(stderr, "Error ejecutando: %s\n", cmd);
    }
}

void particionar_usb(const char *dispositivo) {
    char comando[256];

    // Crear tabla de particiones GPT
    snprintf(comando, sizeof(comando), "parted %s mklabel gpt", dispositivo);
    ejecutar_comando(comando);

    // Crear particiones
    snprintf(comando, sizeof(comando), "parted %s mkpart primary fat32 1MiB 100MiB", dispositivo);
    ejecutar_comando(comando);
    snprintf(comando, sizeof(comando), "parted %s mkpart primary ext4 100MiB 50%%", dispositivo);
    ejecutar_comando(comando);
    snprintf(comando, sizeof(comando), "parted %s mkpart primary ext4 50%% 100%%", dispositivo);
    ejecutar_comando(comando);

    // Formatear partición EFI
    snprintf(comando, sizeof(comando), "mkfs.fat -F32 %s1", dispositivo);
    ejecutar_comando(comando);

    printf("Particionado completado en el dispositivo: %s\n", dispositivo);
}

void escribir_iso(const char *iso_path, const char *particion) {
    char comando[256];

    snprintf(comando, sizeof(comando), "dd if=%s of=%s bs=4M status=progress oflag=sync", iso_path, particion);
    ejecutar_comando(comando);

    printf("ISO escrita en la partición: %s\n", particion);
}

void configurar_grub(const char *dispositivo) {
    char comando[256];

    snprintf(comando, sizeof(comando), "mount %s1 /mnt", dispositivo);
    ejecutar_comando(comando);

    ejecutar_comando("grub-install --target=x86_64-efi --efi-directory=/mnt --boot-directory=/mnt/boot --removable");

    ejecutar_comando("umount /mnt");

    printf("GRUB instalado correctamente en: %s\n", dispositivo);
}

void on_particionar_usb(GtkWidget *widget, gpointer data) {
    const char *dispositivo = (const char *)data;
    particionar_usb(dispositivo);
}

void on_escribir_iso(GtkWidget *widget, gpointer data) {
    const char **params = (const char **)data;
    escribir_iso(params[0], params[1]);
}

void on_configurar_grub(GtkWidget *widget, gpointer data) {
    const char *dispositivo = (const char *)data;
    configurar_grub(dispositivo);
}

int main(int argc, char *argv[]) {
    gtk_init(&argc, &argv);

    GtkWidget *window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title(GTK_WINDOW(window), "Creador de USB Multiboot");
    gtk_window_set_default_size(GTK_WINDOW(window), 400, 300);
    g_signal_connect(window, "destroy", G_CALLBACK(gtk_main_quit), NULL);

    GtkWidget *vbox = gtk_box_new(GTK_ORIENTATION_VERTICAL, 10);
    gtk_container_add(GTK_CONTAINER(window), vbox);

    GtkWidget *btn_particionar = gtk_button_new_with_label("Particionar USB");
    GtkWidget *btn_escribir_iso1 = gtk_button_new_with_label("Escribir Primera ISO");
    GtkWidget *btn_escribir_iso2 = gtk_button_new_with_label("Escribir Segunda ISO");
    GtkWidget *btn_configurar_grub = gtk_button_new_with_label("Instalar GRUB");

    const char *dispositivo = "/dev/sdb"; // Modifica según tu dispositivo USB
    const char *iso1_path = "path/to/iso1.iso"; // Cambiar por la ruta real
    const char *iso2_path = "path/to/iso2.iso"; // Cambiar por la ruta real
    const char *particion1 = "/dev/sdb2";
    const char *particion2 = "/dev/sdb3";

    g_signal_connect(btn_particionar, "clicked", G_CALLBACK(on_particionar_usb), (gpointer)dispositivo);
    g_signal_connect(btn_escribir_iso1, "clicked", G_CALLBACK(on_escribir_iso), (gpointer)(const char *[]){iso1_path, particion1});
    g_signal_connect(btn_escribir_iso2, "clicked", G_CALLBACK(on_escribir_iso), (gpointer)(const char *[]){iso2_path, particion2});
    g_signal_connect(btn_configurar_grub, "clicked", G_CALLBACK(on_configurar_grub), (gpointer)dispositivo);

    gtk_box_pack_start(GTK_BOX(vbox), btn_particionar, FALSE, FALSE, 5);
    gtk_box_pack_start(GTK_BOX(vbox), btn_escribir_iso1, FALSE, FALSE, 5);
    gtk_box_pack_start(GTK_BOX(vbox), btn_escribir_iso2, FALSE, FALSE, 5);
    gtk_box_pack_start(GTK_BOX(vbox), btn_configurar_grub, FALSE, FALSE, 5);

    gtk_widget_show_all(window);
    gtk_main();

    return 0;
}

File: usr/share/installer/installer.py
import subprocess
import os
from tkinter import Tk, Label, Button, filedialog, messagebox, StringVar
from tkinter.ttk import Progressbar

def ejecutar_comando(cmd, progress_bar=None):
    """Ejecuta un comando en el sistema con una barra de progreso opcional."""
    try:
        process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        while process.poll() is None:
            if progress_bar:
                progress_bar.step(5)
        if process.returncode != 0:
            error = process.stderr.read().decode()
            raise RuntimeError(f"Error: {error}")
    except Exception as e:
        raise RuntimeError(f"Error ejecutando '{cmd}': {e}")

def listar_dispositivos_usb():
    """Obtiene una lista de dispositivos USB disponibles."""
    try:
        salida = subprocess.check_output("lsblk -lp | grep 'disk $' | awk '{print $1}'", shell=True)
        dispositivos = salida.decode().strip().split('\n')
        return dispositivos
    except subprocess.CalledProcessError:
        return []

def seleccionar_iso():
    """Permite seleccionar un archivo ISO mediante un cuadro de diálogo."""
    return filedialog.askopenfilename(filetypes=[("Archivos ISO", "*.iso")])

def particionar_usb(dispositivo, progress_bar):
    """Particiona la USB en tres particiones (EFI, ISO 1 e ISO 2)."""
    try:
        comandos = [
            f"parted {dispositivo} mklabel gpt",
            f"parted {dispositivo} mkpart primary fat32 1MiB 100MiB",
            f"parted {dispositivo} mkpart primary ext4 100MiB 50%",
            f"parted {dispositivo} mkpart primary ext4 50% 100%",
            f"mkfs.fat -F32 {dispositivo}1"
        ]
        for cmd in comandos:
            ejecutar_comando(cmd, progress_bar)
        messagebox.showinfo("Éxito", "¡Particionado completado!")
    except Exception as e:
        messagebox.showerror("Error", f"Fallo en el particionado: {e}")

def escribir_iso(iso_path, particion, progress_bar):
    """Escribe una ISO en una partición específica."""
    try:
        ejecutar_comando(f"sudo dd if={iso_path} of={particion} bs=4M status=progress oflag=sync", progress_bar)
        messagebox.showinfo("Éxito", f"¡ISO escrita en {particion}!")
    except Exception as e:
        messagebox.showerror("Error", f"No se pudo escribir la ISO: {e}")

def configurar_grub(dispositivo, progress_bar):
    """Configura GRUB en la partición EFI."""
    try:
        comandos = [
            f"mount {dispositivo}1 /mnt",
            "grub-install --target=x86_64-efi --efi-directory=/mnt --boot-directory=/mnt/boot --removable",
            "umount /mnt"
        ]
        for cmd in comandos:
            ejecutar_comando(cmd, progress_bar)
        messagebox.showinfo("Éxito", "¡GRUB instalado correctamente!")
    except Exception as e:
        messagebox.showerror("Error", f"No se pudo instalar GRUB: {e}")

def main():
    root = Tk()
    root.title("Creador de USB Multiboot")

    dispositivo_seleccionado = StringVar(value="")

    Label(root, text="1. Seleccione el dispositivo USB:").pack()
    dispositivos_usb = listar_dispositivos_usb()
    if dispositivos_usb:
        dispositivo_seleccionado.set(dispositivos_usb[0])
        Label(root, text=f"Dispositivo detectado: {dispositivos_usb[0]}").pack()
    else:
        messagebox.showerror("Error", "¡No se encontraron dispositivos USB!")
        root.quit()

    Label(root, text="2. Seleccione la primera ISO:").pack()
    ruta_iso1 = Button(root, text="Seleccionar ISO 1", command=lambda: seleccionar_iso()).pack()

    Label(root, text="3. Seleccione la segunda ISO:").pack()
    ruta_iso2 = Button(root, text="Seleccionar ISO 2", command=lambda: seleccionar_iso()).pack()

    progress_bar = Progressbar(root, orient="horizontal", mode="determinate", length=300)
    progress_bar.pack()

    Button(root, text="Particionar USB", command=lambda: particionar_usb(dispositivo_seleccionado.get(), progress_bar)).pack()
    Button(root, text="Escribir primera ISO", command=lambda: escribir_iso(ruta_iso1, dispositivo_seleccionado.get() + "2", progress_bar)).pack()
    Button(root, text="Escribir segunda ISO", command=lambda: escribir_iso(ruta_iso2, dispositivo_seleccionado.get() + "3", progress_bar)).pack()
    Button(root, text="Instalar GRUB", command=lambda: configurar_grub(dispositivo_seleccionado.get(), progress_bar)).pack()

    root.mainloop()

if __name__ == "__main__":
    main()
